ime/app/src/main/java/com/menny/android/anysoftkeyboard/AnyApplication.java

```diff
@@ -202,6 +202,13 @@ public void onCreate() {
 
         mPublicNotices = new ArrayList<>(EasterEggs.create());
         mPublicNotices.addAll(Notices.create(this));
+        
+        // For Android 11 (maybe earlier?) we need to make sure the external application directory
+        // exists, to do so, just have the system get an empty file from it and Android will create
+        // it for us.  Likewise we have to do it here instead of in the getBackupFile function as
+        // we don't have the right context by that time.  We need this to write backups too and 
+        // with Android 11 we can no longer do this ourselves.
+        this.getExternalFilesDir("");
```

Comment:
We might not have the `android.permission.READ_EXTERNAL_STORAGE` granted at this time.

Can you share more information about this issue?

Reply:
Sure, according to the note in [this section ](https://developer.android.com/training/data-storage/app-specific#external-access-files) of the developers docs:

```Note: On Android 11 (API level 30) and higher, apps cannot create their own app-specific directory on external storage.```

So when a backup is run and it gets to the `getBackupFile()` call that runs:

```
return new File(
                new File(externalFolder, "/Android/data/" + BuildConfig.APPLICATION_ID + "/files/"),
                filename);
```
The APPLICATION_ID/files directory does not yet exist.

If you try and create it with `.mkdirs()` it fails as Android 11 will no longer let you create the APPLICATION_ID directory in '/Android/data'.   This then causes the backup to fail as it can't write to the non-existant directory.

Using `this.getExternalFilesDir("");` instead of `File(externalFolder, "/Android/data/" + BuildConfig.APPLICATION_ID + "/files/").mkdirs()` forces the system itself to create the directory during the class creation.

It could be moved in to the `getBackupFile()` function instead if `Context` was available but this does not appear to be the case.  I'm not that familiar with Java so perhaps there is a way to get a app context inside this function and then `getExternalFilesDir()` could be called only when the backup actually happens.

I did comple and test the new code, so permissions are available to do the creation (I suspect they aren't needed though as `getExternalFilesDir()` is using something other than the app context to do the directory creation otherwise it would fail just as `mkdirs()` does.

Reply:
thank you!